### PR TITLE
[Clang] Add __builtin_vectorelements to get number of elements in vector

### PR DESCRIPTION
**Pull request by lawben**
Adds a new `__builtin_vectorelements()` function which returns the number of elements for a given vector either at compile-time for fixed-sized vectors, e.g., created via `__attribute__((vector_size(N)))` or at runtime via a call to `@llvm.vscale.i32()` for scalable vectors, e.g., SVE or RISCV V.

The new builtin follows a similar path as `sizeof()`, as it essentially does the same thing but for the number of elements in vector instead of the number of bytes. This allows us to re-use a lot of the existing logic to handle types etc. 

A small side addition is `Type::isSizelessVectorType()`, which we need to distinguish between sizeless vectors (SVE, RISCV V) and sizeless types (WASM). 

This is the [corresponding discussion](https://discourse.llvm.org/t/new-builtin-function-to-get-number-of-lanes-in-simd-vectors/73911).

### CODE REVIEW

**Comment by lawben** [comment]
@erichkeane This is my first PR to the frontend side of LLVM. Please let me know if there is something missing for a Clang PR that I should add.

File: clang/include/clang/AST/Type.h [hunk 1]
```diff
@@ -5456,9 +5459,8 @@ class DeducedTemplateSpecializationType : public DeducedType,
 /// TemplateArguments, followed by a QualType representing the
 /// non-canonical aliased type when the template is a type alias
 /// template.
-class alignas(8) TemplateSpecializationType
-    : public Type,
-      public llvm::FoldingSetNode {
+class alignas(8) TemplateSpecializationType : public Type,
```
Comment (erichkeane) [inline clang/include/clang/AST/Type.h]: Unrelated change?  Or am I missing something?

File: clang/include/clang/Basic/DiagnosticSemaKinds.td [hunk 2]
```diff
@@ -10136,6 +10136,9 @@ def err_vec_builtin_incompatible_vector : Error<
 def err_vsx_builtin_nonconstant_argument : Error<
   "argument %0 to %1 must be a 2-bit unsigned literal (i.e. 0, 1, 2 or 3)">;
 
+def err_vectorelements_non_vector : Error<
+ "'__builtin_vectorelements' argument must be a vector">;
```
Comment (erichkeane) [inline clang/include/clang/Basic/DiagnosticSemaKinds.td:10140]: Do we use similar language elsewhere?  "Must be a vector" is pretty confusing.

File: clang/lib/AST/ExprConstant.cpp [hunk 3]
```diff
@@ -13595,6 +13595,15 @@ bool IntExprEvaluator::VisitUnaryExprOrTypeTraitExpr(
...
             .getQuantity(),
         E);
+  case UETT_VectorElements: {
+    QualType Ty = E->getTypeOfArgument();
+    // If the vector has a fixed size, we can determine the number of elements
+    // at compile time.
+    if (Ty->isVectorType())
+      return Success(Ty->castAs<VectorType>()->getNumElements(), E);
+
+    return false;
```
Comment (erichkeane) [inline clang/lib/AST/ExprConstant.cpp:13605]: Shouldn't this 'return false' just be an  assert?  (That is, instead of the 'if')?  Or is there a way for a non-vector type to make it through?

Additionally, what about sizeless vectors here?  Should we be testing for those instead to make this a 'fail'?  And does that case need a diagnostic?

File: clang/lib/CodeGen/CGExprScalar.cpp [hunk 4]
```diff
@@ -3083,6 +3083,19 @@ ScalarExprEmitter::VisitUnaryExprOrTypeTraitExpr(
                 E->getTypeOfArgument()->getPointeeType()))
             .getQuantity();
     return llvm::ConstantInt::get(CGF.SizeTy, Alignment);
+  } else if (E->getKind() == UETT_VectorElements) {
+    // For scalable vectors, we don't know the size at compile time. We can use
+    // @llvm.vscale to calculate it at runtime.
+    if (E->getTypeOfArgument()->isSizelessVectorType()) {
```
Comment (erichkeane) [inline clang/lib/CodeGen/CGExprScalar.cpp:3089]: Where do you handle non-sizeless vector types that make it here?

File: clang/lib/CodeGen/CGExprScalar.cpp [hunk 5]
```diff
@@ -3083,6 +3083,19 @@ ScalarExprEmitter::VisitUnaryExprOrTypeTraitExpr(
...
+      auto *VecTy = dyn_cast<llvm::ScalableVectorType>(
+          ConvertType(E->getTypeOfArgument()));
+      uint64_t NumUnscaledElements = VecTy->getMinNumElements();
```
Comment (erichkeane) [inline clang/lib/CodeGen/CGExprScalar.cpp:3092]: I'm not particularly familiar with the sizeless vector type here, I would have expected the argument here to be an unknown value/expression, right?  Or is VScale doing something I'm unfamiliar with?  But it too doesn't seem to take the vector type as an argument.

File: clang/lib/Sema/SemaExpr.cpp [hunk 6]
```diff
@@ -4351,6 +4352,17 @@ static bool CheckVecStepTraitOperandType(Sema &S, QualType T,
   return false;
 }
 
+static bool CheckVectorElementsTraitOperandType(Sema &S, QualType T,
+                                                SourceLocation Loc,
+                                                SourceRange ArgRange) {
+  // builtin_vectorelements supports both fixed-sized and scalable vectors.
+  if (!T->isVectorType() && !T->isSizelessVectorType()) {
+    S.Diag(Loc, diag::err_vectorelements_non_vector) << T << ArgRange;
```
Comment (erichkeane) [inline clang/lib/Sema/SemaExpr.cpp:4360]: remove the curleys and just do: `return S.Diag`

File: clang/test/Sema/builtin_vectorelements.c [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+  (void) __builtin_vectorelements(veci4);
+
+  veci4 vec;
+  (void) __builtin_vectorelements(vec);
+
+  typedef veci4 some_other_vec;
+  (void) __builtin_vectorelements(some_other_vec);
+
+  struct Foo { int a; };
+  __builtin_vectorelements(struct Foo); // expected-error {{'__builtin_vectorelements' argument must be a vector}}
```
Comment (erichkeane) [inline clang/test/Sema/builtin_vectorelements.c:21]: I would like to see some constexpr tests as well.

File: clang/test/CodeGen/builtin_vectorelements.c [hunk 8]
```diff
@@ -0,0 +1,121 @@
+// RUN: %clang_cc1 -O1 -triple aarch64 -target-feature +neon %s -emit-llvm -o - | FileCheck --check-prefixes=CHECK,NEON %s
```
Comment (erichkeane) [inline clang/test/CodeGen/builtin_vectorelements.c:1]: Add -disable-llvm-passes to all of these.

**Comment by erichkeane** [comment]
Also, needs a release note.

File: clang/include/clang/AST/Type.h (hunk 1 above)
Comment (lawben) [reply clang/include/clang/AST/Type.h]: This was a clang-format error because of a merge conflict. Reverted it.

File: clang/include/clang/Basic/DiagnosticSemaKinds.td (hunk 2 above)
Comment (lawben) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td:10140]: this was inspired by "`first argument to __builtin_convertvector must be a vector`". but if you have a suggestion, i'm happy to change this.

Comment (erichkeane) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td:10140]: Actually, we probably want to combine those two with a 'select' or insertion instead of keeping them as separate.  I might do something like `must be of vector type` instead of `a vector`, unless there is something more obviously specific (I think `SIMD vector type` would be incorrect, right?)

File: clang/lib/AST/ExprConstant.cpp (hunk 3 above)
Comment (lawben) [reply clang/lib/AST/ExprConstant.cpp:13605]: maybe my understanding here is wrong, so please correct me in that case. If the vector is fixed-sized, we can determine the number of elements at compile time, which is why we return the number here in `ExprConstant.cpp`. if it is a scalable vector, we `return false`, as we cannot determine a constant value. in that case, we move along and hit `CGExprScalar.cpp`, which is where we emit the runtime `vscale` code for anything that is not constant. 

So in my understanding, we either emit a constant in `ExprConstant` or the runtime logic in `CGExprScalar`. I can add assertions to check that we are always in a good case, but i've not come across a fixed-sized vector path that does not get caught in `ExprConstant`. if there are alternative code paths, then you are correct and I should add checks. but I don't want to duplicate logic if this case can never happen anyway.

File: clang/lib/CodeGen/CGExprScalar.cpp (hunk 5 above)
Comment (lawben) [reply clang/lib/CodeGen/CGExprScalar.cpp:3092]: the argument here is the vector type that was passed to `__builtin_vectorelements()`. the call to `@llvm.vscale` is independent of the type, as it is a system-wide thing, so it does not need the type. 

scalable vectors are represented as `<vscale x NUM_ELEMENTS_PER_MINIMUM_VECTOR x TYPE>`. so if we get `vscale`, we still need to figure out the `NUM_ELEMENTS` part so we can get the actual number of elements on this system by multiplying it with. as that depends on the vector type passed to the builtin, as the number of elements depends on the element's type. for SVE, LLVM assumes a minimum 16-byte vector, so a `<vscale x 4 x i32`> has `4 * vscale` but a `<vscale x 8 x i16>` has `8 * vscale`. 

I hope this clarifies the code. if so, I'll probably add a comment to explain.

File: clang/lib/CodeGen/CGExprScalar.cpp [hunk 9]
```diff
@@ -3083,6 +3083,19 @@ ScalarExprEmitter::VisitUnaryExprOrTypeTraitExpr(
...
+
+      llvm::Value *VScale =
+          Builder.CreateVScale(llvm::ConstantInt::get(CGF.SizeTy, 1));
+      return Builder.CreateMul(
+          VScale, llvm::ConstantInt::get(CGF.SizeTy, NumUnscaledElements));
```
Comment (nikic) [inline clang/lib/CodeGen/CGExprScalar.cpp:3097]: You can use `Builder.CreateElementCount(CGF.SizeTy, VecTy->getElementCount())` or so here. (Incidentally, this works for both fixed and scalable vectors.)

File: clang/lib/AST/ExprConstant.cpp (hunk 3 above)
Comment (erichkeane) [reply clang/lib/AST/ExprConstant.cpp:13605]: Your understanding of how this works is correct, I missed you were handling sizeless vectors that way.  So above 'return false' there 
could be an `assert(Ty->isSizelessVectorType())` ?  

Note that we're not guaranteed to come to ExprConstant.cpp in the case where it is a sized-vector, so you likely need to make sure the CGExprScalar code handles sized types as well.

File: clang/lib/CodeGen/CGExprScalar.cpp (hunk 9 above)
Comment (lawben) [reply clang/lib/CodeGen/CGExprScalar.cpp]: Neat, thanks for the pointer. that is a lot nicer. @erichkeane this should fix the comments above, as we now handle both types of vectors here and the code is more clear.

File: clang/test/Sema/builtin_vectorelements.c (hunk 7 above)
Comment (lawben) [reply clang/test/Sema/builtin_vectorelements.c]: Are you fine with this giving an error like: (this is what is currently thrown)

```
constexpr variable 'foo' must be initialized by a constant expression
```

or should we explicitly handle this somewhere to say something like: 

```
__builtin_vectorelements cannot be consexpr evaluated with scalable vectors
```

(PS: I'm not really sure how/where to check this for the second case)

Comment (erichkeane) [reply clang/test/Sema/builtin_vectorelements.c]: I've not spent much time lately in the constexpr evaluator, but I believe there is a way to just add a note to that, which you should do, I THINK it is something like `FFDiag`

Comment (lawben) [reply clang/test/Sema/builtin_vectorelements.c]: just FYI: its `CCEDiag`

**Comment by lawben** [comment]
@erichkeane I think I've addressed all of your comments so far. Please check if there is anything else missing.

File: clang/lib/AST/ItaniumMangle.cpp [hunk 10]
```diff
@@ -5126,6 +5126,14 @@ void CXXNameMangler::mangleExpression(const Expr *E, unsigned Arity,
       Diags.Report(DiagID);
       return;
     }
+    case UETT_VectorElements: {
```
Comment (erichkeane) [inline clang/lib/AST/ItaniumMangle.cpp:5129]: We probably need the equivilent here for the MicrosoftMangle.cpp as well?

File: clang/lib/CodeGen/CGExprScalar.cpp [hunk 11]
```diff
@@ -3083,6 +3083,10 @@ ScalarExprEmitter::VisitUnaryExprOrTypeTraitExpr(
                 E->getTypeOfArgument()->getPointeeType()))
             .getQuantity();
     return llvm::ConstantInt::get(CGF.SizeTy, Alignment);
+  } else if (E->getKind() == UETT_VectorElements) {
+    auto *VecTy =
+        dyn_cast<llvm::VectorType>(ConvertType(E->getTypeOfArgument()));
```
Comment (erichkeane) [inline clang/lib/CodeGen/CGExprScalar.cpp:3088]:
```suggestion
        cast<llvm::VectorType>(ConvertType(E->getTypeOfArgument()));
```
You immediately dereference this anyway, and I think the semantics guarantees this is the case, correct?

**Review by erichkeane — COMMENTED** [review]
2 more nits, else LGTM.

File: clang/lib/AST/ItaniumMangle.cpp (hunk 10 above)
Comment (lawben) [reply clang/lib/AST/ItaniumMangle.cpp:5129]: I don't know. But none of the other `UETT_*` types are covered in `MicrosoftMangle.cpp`. I'm not sure whether they are missing or if it not necessary to add this.

Comment (erichkeane) [reply clang/lib/AST/ItaniumMangle.cpp:5129]: OK, SGTM at least, I just know this pattern of 'cannot yet mangle...' shows up primarily in the MicrosoftMangle, so figured it might need to be covered.  There is some funny-business as to how much each mangles of expressions in template arguments.

This gets me thinking further though,the constexprness of this likely means you may need to mangle this.  I'm open if others are to letting it be done in a follow-up however.

**Review by erichkeane — APPROVED** [review]
I'm good with it if the pre-commit is, and no one comes out and really needs the mangling done in this patch.

**Comment by lawben** [comment]
As nobody has come forward in the the last two days, I'm gonna merge this now. I'll have a look a mangling this for Microsoft in a follow-up, as this may require a few changes. I'm not yet sure what has to be changed for mangling.

**Comment by tbaederr** [comment]
Have you seen the failing buildbots?

**Comment by lawben** [comment]
@tbaederr Jupp, I'm on it. I was not aware that I had to add `REQUIRES` to all the tests for it to find the header files. I assumed they are always available. See #69582.

File: clang/lib/AST/ItaniumMangle.cpp (hunk 10 above)
Comment (lawben) [reply clang/lib/AST/ItaniumMangle.cpp:5129]: @philnik777 I'm adding you here because of your recent commit that adds `__datasizeof` (#67805). This is essentially the same code path and you also specify the "cannot mangle" for Itanium but not for Microsoft. Do you know if this is needed for Microsoft (see comments above)? 

I'm happy to address this for `__builtin_vectorelements`, but I really don't know where to start and `__datasizeof` seems to be the same.
